test/tests_basic.jl

```diff
@@ -20,6 +20,34 @@ function iter_volume_for_p4est_qcoord_to_vertex(info::Ptr{p4est_iter_volume_info
   return nothing
 end
 
+# These functions belong to the testset "p4est_refine_coarsen" below
+function refine_fn(p4est, which_tree, quadrant)
+  quadrant_obj = unsafe_load(quadrant)
+  if quadrant_obj.level < 2
+    return Cint(1)
+  else
+    return Cint(0)
+  end
+end
+
+function coarsen_fn(p4est, which_tree, quadrant)
+  quadrant_obj = unsafe_load(quadrant)
+  if quadrant_obj.level > 0
+    return Cint(1)
+  else
+    return Cint(0)
+  end
+end
```

Comment:
In C, this has to be of type `p4est_refine_t`:
```julia
help?> p4est_coarsen_t
search: p4est_coarsen_t p4est_coarsen_ext p4est_coarsen p4est_partition_for_coarsening

  Callback function prototype to decide for coarsening.

  Parameters
  ––––––––––––

    •  p4est:[in] the forest

    •  which_tree:[in] the tree containing quadrant

    •  quadrants:[in] Pointers to 4 siblings in Morton ordering.

  Returns
  –––––––––

  nonzero if the quadrants shall be replaced with their parent.
```
Thus, you do not have a single `quadrant` as last argument but a C style array of pointers to quadrants. I found this while changing this function to
```julia
julia> function coarsen_fn(p4est, which_tree, quadrant)
         quadrant_obj = unsafe_load(quadrant)
         @info "coarsen_fn" unsafe_load(quadrant.x) unsafe_load(quadrant.y) unsafe_load(quadrant.level)
         if quadrant_obj.level > 0
           return Cint(1)
         else
           return Cint(0)
         end
       end
```
Running the test below gives garbage, e.g.,
```
┌ Info: coarsen_fn
│   quadrant = Ptr{p4est_quadrant} @0x00007ffd674ba650
│   unsafe_load(quadrant.x) = 61432064
│   unsafe_load(quadrant.y) = 0
└   unsafe_load(quadrant.level) = 24
┌ Info: coarsen_fn
│   quadrant = Ptr{p4est_quadrant} @0x00007ffd674ba650
│   unsafe_load(quadrant.x) = 61432088
│   unsafe_load(quadrant.y) = 0
└   unsafe_load(quadrant.level) = 120
┌ Info: coarsen_fn
│   quadrant = Ptr{p4est_quadrant} @0x00007ffd674ba650
│   unsafe_load(quadrant.x) = 61432112
│   unsafe_load(quadrant.y) = 0
└   unsafe_load(quadrant.level) = -40
┌ Info: coarsen_fn
│   quadrant = Ptr{p4est_quadrant} @0x00007ffd674ba650
│   unsafe_load(quadrant.x) = 61432208
│   unsafe_load(quadrant.y) = 0
└   unsafe_load(quadrant.level) = 56
```
This is because the last argument of `coarsen_fn_c` below should be a `Ptr{Ptr{p4est_quadrant_t}}`, not a `Ptr{p4est_quadrant_t}`.

Reply:
You are right. I changed the coarsen_fn now to just return Cint(1), which should do the job.